docs(blog): google app actions integration, pt 3 (#130)
spokestack/spokestack-website: content/blog/integrating-spokestack-google-app-actions-3/index.md

---
title: Integrating Spokestack with Google App Actions, Part 3
date: '2020-12-10'

> description: Add a UI to your app's voice features using Spokestack Tray

Reviewer: i am confused by this description

Author: Spokestack gives your app voice support, but there's no UI to give users feedback about that voice support without Tray (or something custom). Do you have an idea for how to make it clearer? Should it just be longer?

author: josh
tags: Android, Engineering, Tutorial
draft: false
---

_This tutorial is part of a series:_

- [Part 1](/blog/integrating-spokestack-google-app-actions-1): Working with Google App Actions
- [Part 2](/blog/integrating-spokestack-google-app-actions-2): Adding your own voice experience with Spokestack
- **Part 3** (_You are here!_): Using Spokestack Tray to add a voice UI

---

Now that we know what we need to get started with voice control, both on Google's end and in our app, we'll polish our user experience a bit by adding visual elements to give the user more feedback about what the app is hearing and saying. With the changes we'll make in this tutorial, an interaction with Google Assistant will look like this:

![screen capture of Google Assistant handing voice control to an app](./images/app-actions-demo.gif)

To follow along, run `git checkout spokestack-tray` in your copy of the [sample repository](https://github.com/spokestack/app-actions-example). Since this tutorial involves changes to the activities we've already introduced, we've made them on a separate branch for easy comparison to the "headless" version of Spokestack.

## The why and how of Tray

> [Spokestack Tray](https://github.com/spokestack/spokestack-tray-android) exists to help you visualize voice interactions in your app without building an entire UI for it from scratch. It includes a microphone button used to activate ASR without the use of a wakeword, visual feedback while ASR is active, and a chat-like interface that displays the conversation between user and app.

Reviewer: ```suggestion
[Spokestack Tray](https://github.com/spokestack/spokestack-tray-android) exists to help you easily add voice interactions in your app without building an entire UI for it from scratch. It includes a microphone button used to activate ASR without the use of a wakeword, visual feedback while ASR is active, and a chat-like interface that displays the conversation between user and app.
```

Author: You _add_ the voice interactions by just using Spokestack; the only thing Spokestack Tray adds to the picture is the visualization (and with it, some management of Spokestack features).

> Let's walk through the process of adding it to our project, starting with the changes to `build.gradle`. Since the Tray exists to mediate interaction with Spokestack, it bundles the dependencies necessary for that, so we can replace all our Spokestack-related dependencies with the following:

Reviewer: ```suggestion
Let's walk through the process of dropping it to our project, starting with the changes to `build.gradle`. Since the Tray exists to mediate your users interaction with Spokestack, it bundles the dependencies necessary for that. Hence we can replace all our Spokestack-related dependencies with the following:
```

```groovy
    implementation 'io.spokestack:tray:0.3.0'
```

Continuing with the declarative parts of the project, Tray is implemented as a `Fragemnt`, so each `Activity` that wants to include it will want to add a few lines to its layout XML:

```xml
    <!-- nested in the main layout, after other views/sublayouts -->

    <include
        android:id="@+id/tray_fragment"
        layout="@layout/spokestack_tray_fragment"
        />
```

This will add a nested `ConstraintLayout` containing the `SpokestackTray` fragment into your layout. If you'd rather manage the layout XML yourself, take a look at [the layout file](https://github.com/spokestack/spokestack-tray-android/blob/main/SpokestackTray/src/main/res/layout/spokestack_tray_fragment.xml) we're including. You'll want to use the same tag for the fragment as it does, as it's important elsewhere.

Note also the `clipChildren` attribute on the parent layout. Depending on how your app is set up, you might also need to add this to any layouts serving as parents to the tray fragment. Our sample app does this in all its activities.

## On to the fun stuff!

> That takes care of the setup, so let's move on to the code from [part 2](/blog/integrating-spokestack-google-app-actions-2) that we'll need to change. Spoiler alert: We'll be deleting more than we add.

Reviewer: ```suggestion
That takes care of the setup, so let's move on to the code from [part 2](/blog/integrating-spokestack-google-app-actions-2) that we'll need to change. Spoiler alert: thanks to Spokestack Tray's magic, we'll be deleting more than we add, resulting in a much simpler developer experience!
```

Author: Ha, maybe a bit bombastic, but point taken.
